src/collapse/dtrace.rs

```diff
@@ -0,0 +1,214 @@
+use super::Collapse;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::{HashMap, VecDeque};
+use std::io;
+use std::io::prelude::*;
+
+const HEADER_LINES: u8 = 3;
+
+lazy_static! {
+    static ref CPP_REGEX: Regex = Regex::new("(?P<n>::.*)[(<].*").unwrap();
+}
+
+/// Settings that change how frames are named from the incoming stack traces.
+///
+/// All options default to off.
+#[derive(Clone, Debug, Default)]
+pub struct Options {
+    /// include function offset (except leafs)
+    pub includeoffset: bool,
+}
+
+/// A stack collapser for the output of dtrace `ustrace()`.
+///
+/// To construct one, either use `dtrace::Folder::default()` or create an [`Options`] and use
+/// `dtrace::Folder::from(options)`.
+#[derive(Default)]
+pub struct Folder {
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Keep track of stack string size while we consume a stack
+    stack_str_size: usize,
+
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        for _ in 0..HEADER_LINES {
+            reader.read_line(&mut line)?;
+        }
```

Comment:
I wonder if we could even get rid of the constant here with:
```rust
// skip header lines -- first empty line marks start of data
loop {
    line.clear();
    reader.read_line(&mut line)?;
    if line.is_empty() {
        break;
    }
}
```

Reply:
You may also technically need to check for `read_line() == 0` here.

Reply:
👍